src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php

```diff
@@ -91,12 +90,12 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 $form = $event->getForm();
                 $data = $event->getData();
 
+                // Since the type always use mapper an empty array will not be
+                // considered as empty in Form::submit(), we need to evaluate
+                // empty data here so its value is submitted to sub forms
                 if (null === $data) {
                     $emptyData = $form->getConfig()->getEmptyData();
-
-                    if (false === FormUtil::isEmpty($emptyData) && array() !== $emptyData) {
-                        $data = is_callable($emptyData) ? call_user_func($emptyData, $form, $data) : $emptyData;
-                    }
+                    $data = $emptyData instanceof \Closure ? $emptyData($form, $data) : $emptyData;
```

Comment:
is `instanceof \Closure` legit instead of `is_callable` (same below?)

Reply:
This is totally legit and what must be, this is the point of this bug fix. Only a `\Closure` should be considered here, nothing else.

The view data can be:
- an object (whose class should be the same as defined by the `data_class` option
- an array
- scalar

The three of them can be callable:
- the object can be invokable
- the array can be a class or an object with a method name
- the string a function name (which triggered the related issue)

Deprecating a broken behavior on master does not make any sense to me,
I really think we need to merge this PR as a bug fix, because the [previous one](#17822) was wrong.